File: android/app/src/main/java/com/readypackets/mobile/core/AuthCoordinator.kt
package com.readypackets.mobile.core

import android.content.Context
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import com.readypackets.mobile.BuildConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class AuthCoordinator(
    private val context: Context,
    private val api: MobileApiClient,
    private val tokenStore: SecureTokenStore,
    private val onAuthenticated: () -> Unit,
    private val onFailure: (String) -> Unit,
) {
    private var state: String? = null
    private var verifier: String? = null
    private val installationId = context.getSharedPreferences("readypackets_installation", Context.MODE_PRIVATE)
        .getString("id", null)
        ?: java.util.UUID.randomUUID().toString().lowercase().also {
            context.getSharedPreferences("readypackets_installation", Context.MODE_PRIVATE)
                .edit().putString("id", it).apply()
        }

    fun begin() {
        val newState = Pkce.state()
        val newVerifier = Pkce.verifier()
        state = newState
        verifier = newVerifier
        tokenStore.saveAuthorizationAttempt(newState, newVerifier)
        val uri = Uri.parse(BuildConfig.PORTAL_BASE_URL.trimEnd('/') + "/api/mobile/v1/authorize")
            .buildUpon()
            .appendQueryParameter("client_id", "readypackets-native")
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("redirect_uri", BuildConfig.OAUTH_REDIRECT_URI)
            .appendQueryParameter("code_challenge", Pkce.challenge(newVerifier))
            .appendQueryParameter("code_challenge_method", "S256")
            .appendQueryParameter("state", newState)
            .appendQueryParameter("device_id", installationId)
            .appendQueryParameter("platform", "android")
            .appendQueryParameter("app_version", BuildConfig.VERSION_NAME)
            .appendQueryParameter("scope", "mobile:read mobile:write")
            .build()
        CustomTabsIntent.Builder().setShowTitle(true).build().launchUrl(context, uri)
    }

    fun complete(uri: Uri) {
        val expectedRedirect = Uri.parse(BuildConfig.OAUTH_REDIRECT_URI)
        val persistedAttempt = tokenStore.validAuthorizationAttempt()
        val expectedState = state ?: persistedAttempt?.first
        val expectedVerifier = verifier ?: persistedAttempt?.second
        if (uri.scheme != expectedRedirect.scheme || uri.host != expectedRedirect.host ||
            uri.path != expectedRedirect.path || uri.getQueryParameter("state") != expectedState
        ) {
            tokenStore.clearAuthorizationAttempt()
            onFailure("The secure sign-in return could not be verified.")
            return
        }
        val code = uri.getQueryParameter("code")
        if (code.isNullOrBlank() || expectedVerifier == null) {
            tokenStore.clearAuthorizationAttempt()
            onFailure("The sign-in request expired. Please try again.")
            return
        }
        CoroutineScope(Dispatchers.Main).launch {
            api.exchangeCode(code, expectedVerifier)
                .onSuccess {
                    state = null
                    verifier = null
                    tokenStore.clearAuthorizationAttempt()
                    onAuthenticated()
                }
                .onFailure { onFailure(it.message ?: "Sign-in failed.") }
        }
    }
}

File: ios/ReadyPackets/ReadyPacketsApp/Core/AuthCoordinator.swift
import Foundation
import UIKit
import AuthenticationServices
import CryptoKit
import Security

@MainActor
final class AuthCoordinator {
    private let api: MobileAPIClient
    private let tokenStore: TokenStore
    private var state: String?
    private var verifier: String?
    private let installationID: String
    private var webAuthenticationSession: ASWebAuthenticationSession?
    private let presentationContext = AuthenticationPresentationContext()

    init(api: MobileAPIClient, tokenStore: TokenStore) {
        self.api = api
        self.tokenStore = tokenStore
        installationID = DeviceInstallation.id
    }

    func begin() async throws {
        let newState = try Self.randomURLSafeString(byteCount: 32)
        let newVerifier = try Self.randomURLSafeString(byteCount: 64)
        try await tokenStore.saveAuthorizationAttempt(state: newState, verifier: newVerifier)
        state = newState
        verifier = newVerifier
        let challenge = Data(SHA256.hash(data: Data(newVerifier.utf8))).base64URLEncodedString()
        var components = URLComponents(
            url: AppConfig.portalBaseURL.appending(path: "/api/mobile/v1/authorize"),
            resolvingAgainstBaseURL: false,
        )!
        components.queryItems = [
            URLQueryItem(name: "client_id", value: "readypackets-native"),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "redirect_uri", value: AppConfig.redirectURI.absoluteString),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: newState),
            URLQueryItem(name: "device_id", value: installationID),
            URLQueryItem(name: "platform", value: "ios"),
            URLQueryItem(name: "app_version", value: AppConfig.appVersion),
            URLQueryItem(name: "scope", value: "mobile:read mobile:write"),
        ]
        guard let url = components.url else { throw URLError(.badURL) }
        let session = ASWebAuthenticationSession(
            url: url,
            callbackURLScheme: AppConfig.redirectURI.scheme,
        ) { [weak self] callback, _ in
            guard let callback else { return }
            Task { await self?.complete(with: callback) }
        }
        session.presentationContextProvider = presentationContext
        session.prefersEphemeralWebBrowserSession = false
        webAuthenticationSession = session
        guard session.start() else {
            await tokenStore.clearAuthorizationAttempt()
            throw APIProblem(
                title: "Secure sign-in unavailable",
                code: "browser_session_failed",
                detail: "The system browser could not start the sign-in session.",
            )
        }
    }

    func complete(with url: URL) async throws {
        guard url.scheme == AppConfig.redirectURI.scheme,
              url.host == AppConfig.redirectURI.host,
              url.path == AppConfig.redirectURI.path else {
            throw APIProblem(
                title: "Unrecognized sign-in return",
                code: "redirect_mismatch",
                detail: "The sign-in callback did not match this application.",
            )
        }
        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        let returnedState = components?.queryItems?.first(where: { $0.name == "state" })?.value
        let attempt = await tokenStore.validAuthorizationAttempt()
        let expectedState = state ?? attempt?.state
        let expectedVerifier = verifier ?? attempt?.verifier
        guard returnedState == expectedState,
              let code = components?.queryItems?.first(where: { $0.name == "code" })?.value,
              let expectedVerifier else {
            await tokenStore.clearAuthorizationAttempt()
            throw APIProblem(
                title: "Sign-in could not be verified",
                code: "state_mismatch",
                detail: "Please begin sign-in again.",
            )
        }
        let tokens = try await api.tokenRequest(form: [
            "grant_type": "authorization_code",
            "client_id": "readypackets-native",
            "code": code,
            "code_verifier": expectedVerifier,
            "redirect_uri": AppConfig.redirectURI.absoluteString,
        ])
        try await tokenStore.save(tokens: tokens)
        await tokenStore.clearAuthorizationAttempt()
        state = nil
        verifier = nil
        webAuthenticationSession = nil
    }

    private static func randomURLSafeString(byteCount: Int) throws -> String {
        var bytes = [UInt8](repeating: 0, count: byteCount)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
            throw APIProblem(
                title: "Secure sign-in unavailable",
                code: "randomness_unavailable",
                detail: "Your device could not prepare a secure authorization request.",
            )
        }
        return Data(bytes).base64URLEncodedString()
    }
}

private final class AuthenticationPresentationContext: NSObject, ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        return scenes.first?.keyWindow ?? ASPresentationAnchor()
    }
}

private extension Data {
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
